File: service/control_plane.py
"""Signed, allowlisted metadata projection from a Diffuse data plane."""

from __future__ import annotations

import hashlib
import hmac
import os
import stat
import time
import urllib.request
from pathlib import Path
from typing import Literal
from urllib.parse import urlsplit

from pydantic import BaseModel, ConfigDict, Field

from service.scm import normalize_base_url


class Projection(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)


class DeploymentProjection(Projection):
    deploymentKey: str = Field(min_length=1, max_length=200)
    name: str = Field(min_length=1, max_length=200)
    kind: Literal["self_hosted", "managed"]
    region: str | None = Field(default=None, max_length=100)
    status: Literal["healthy", "degraded", "offline"]
    version: str = Field(min_length=1, max_length=100)
    updatedAt: int = Field(ge=0)


class RepositoryProjection(Projection):
    externalId: str = Field(min_length=1, max_length=200)
    name: str = Field(min_length=1, max_length=512)
    provider: Literal["github", "gitlab"]
    defaultBranch: str = Field(min_length=1, max_length=255)
    indexStatus: Literal["ready", "indexing", "stale", "failed"]
    reviewStatus: Literal["idle", "reviewing", "attention"]
    lastIndexedAt: int | None = Field(default=None, ge=0)
    lastReviewAt: int | None = Field(default=None, ge=0)
    openFindingCount: int = Field(ge=0)
    criticalFindingCount: int = Field(ge=0)
    updatedAt: int = Field(ge=0)


class ReviewProjection(Projection):
    externalId: str = Field(min_length=1, max_length=200)
    repositoryExternalId: str = Field(min_length=1, max_length=200)
    repositoryName: str = Field(min_length=1, max_length=512)
    number: int | None = Field(default=None, ge=0)
    title: str = Field(min_length=1, max_length=500)
    status: Literal["queued", "reviewing", "published", "failed"]
    findingCount: int = Field(ge=0)
    criticalCount: int = Field(ge=0)
    latencyMs: int | None = Field(default=None, ge=0)
    model: str = Field(min_length=1, max_length=512)
    updatedAt: int = Field(ge=0)


class QualityProjection(Projection):
    evaluatedAt: int = Field(ge=0)
    precision: float = Field(ge=0, le=1)
    recall: float = Field(ge=0, le=1)
    f1: float = Field(ge=0, le=1)
    truePositives: int = Field(ge=0)
    falsePositives: int = Field(ge=0)
    falseNegatives: int = Field(ge=0)
    addressedFindings: int = Field(ge=0)
    sampleSize: int = Field(ge=0)
    medianLatencyMs: int = Field(ge=0)
    estimatedCostUsd: float = Field(ge=0)


class ModelProjection(Projection):
    provider: str = Field(min_length=1, max_length=100)
    model: str = Field(min_length=1, max_length=512)
    status: Literal["ready", "missing_key", "error"]
    reviewPasses: list[str] = Field(max_length=10)
    updatedAt: int = Field(ge=0)


class ControlPlaneSnapshot(Projection):
    deployment: DeploymentProjection
    repositories: list[RepositoryProjection] = Field(max_length=50)
    reviews: list[ReviewProjection] = Field(max_length=40)
    quality: QualityProjection | None = None
    model: ModelProjection | None = None


def _signing_secret(path: Path) -> bytes:
    if path.is_symlink() or not path.is_file():
        raise ValueError("control-plane signing secret must be a regular file")
    mode = stat.S_IMODE(path.stat().st_mode)
    if mode & 0o077:
        raise ValueError("control-plane signing secret must not be group/world accessible")
    secret = path.read_bytes().strip()
    if len(secret) < 32:
        raise ValueError("control-plane signing secret must contain at least 32 bytes")
    return secret


def publish_snapshot(
    snapshot: ControlPlaneSnapshot,
    *,
    url: str | None = None,
    secret_file: Path | None = None,
) -> None:
    base_url = normalize_base_url(
        url or os.environ.get("DIFFUSE_CONTROL_PLANE_URL", ""),
        field_name="DIFFUSE_CONTROL_PLANE_URL",
    )
    parsed = urlsplit(base_url)
    if parsed.scheme != "https" and parsed.hostname not in {"localhost", "127.0.0.1", "::1"}:
        raise ValueError("DIFFUSE_CONTROL_PLANE_URL must use HTTPS off loopback")
    path = secret_file or Path(
        os.environ.get(
            "DIFFUSE_CONTROL_PLANE_SIGNING_SECRET_FILE",
            "/run/secrets/diffuse-control-plane-signing-key",
        )
    )
    secret = _signing_secret(path)
    body = snapshot.model_dump_json(exclude_none=True).encode()
    timestamp = str(int(time.time()))
    signature = hmac.new(
        secret,
        timestamp.encode() + b"." + body,
        hashlib.sha256,
    ).hexdigest()
    request = urllib.request.Request(
        f"{base_url}/v1/data-plane/snapshot",
        data=body,
        method="POST",
        headers={
            "Content-Type": "application/json",
            "X-Diffuse-Timestamp": timestamp,
            "X-Diffuse-Signature": f"sha256={signature}",
        },
    )
    with urllib.request.urlopen(request, timeout=10) as response:
        if response.status != 202:
            raise RuntimeError(f"control plane rejected snapshot with HTTP {response.status}")


def load_snapshot(path: Path) -> ControlPlaneSnapshot:
    if path.is_symlink() or not path.is_file():
        raise ValueError("control-plane snapshot must be a regular file")
    return ControlPlaneSnapshot.model_validate_json(path.read_text())

File: diffuse-control/convex/controlPlane.ts
import { v } from "convex/values";

import { internalMutation, query } from "./_generated/server";

const DEMO_DEPLOYMENT = "demo-local";

const deploymentProjection = v.object({
  deploymentKey: v.string(),
  name: v.string(),
  kind: v.union(v.literal("self_hosted"), v.literal("managed")),
  region: v.optional(v.string()),
  status: v.union(
    v.literal("healthy"),
    v.literal("degraded"),
    v.literal("offline"),
  ),
  version: v.string(),
  updatedAt: v.number(),
});

const repositoryProjection = v.object({
  externalId: v.string(),
  name: v.string(),
  provider: v.union(v.literal("github"), v.literal("gitlab")),
  defaultBranch: v.string(),
  indexStatus: v.union(
    v.literal("ready"),
    v.literal("indexing"),
    v.literal("stale"),
    v.literal("failed"),
  ),
  reviewStatus: v.union(
    v.literal("idle"),
    v.literal("reviewing"),
    v.literal("attention"),
  ),
  lastIndexedAt: v.optional(v.number()),
  lastReviewAt: v.optional(v.number()),
  openFindingCount: v.number(),
  criticalFindingCount: v.number(),
  updatedAt: v.number(),
});

const reviewProjection = v.object({
  externalId: v.string(),
  repositoryExternalId: v.string(),
  repositoryName: v.string(),
  number: v.optional(v.number()),
  title: v.string(),
  status: v.union(
    v.literal("queued"),
    v.literal("reviewing"),
    v.literal("published"),
    v.literal("failed"),
  ),
  findingCount: v.number(),
  criticalCount: v.number(),
  latencyMs: v.optional(v.number()),
  model: v.string(),
  updatedAt: v.number(),
});

const qualityProjection = v.object({
  evaluatedAt: v.number(),
  precision: v.number(),
  recall: v.number(),
  f1: v.number(),
  truePositives: v.number(),
  falsePositives: v.number(),
  falseNegatives: v.number(),
  addressedFindings: v.number(),
  sampleSize: v.number(),
  medianLatencyMs: v.number(),
  estimatedCostUsd: v.number(),
});

const modelProjection = v.object({
  provider: v.string(),
  model: v.string(),
  status: v.union(
    v.literal("ready"),
    v.literal("missing_key"),
    v.literal("error"),
  ),
  reviewPasses: v.array(v.string()),
  updatedAt: v.number(),
});

export const dashboard = query({
  args: {},
  returns: v.object({
    deployment: v.union(
      v.null(),
      deploymentProjection.extend({
        _id: v.id("deployments"),
        _creationTime: v.number(),
      }),
    ),
    repositories: v.array(
      repositoryProjection.extend({
        _id: v.id("repositories"),
        _creationTime: v.number(),
        deploymentKey: v.string(),
      }),
    ),
    reviews: v.array(
      reviewProjection.extend({
        _id: v.id("reviewRuns"),
        _creationTime: v.number(),
        deploymentKey: v.string(),
      }),
    ),
    quality: v.union(
      v.null(),
      qualityProjection.extend({
        _id: v.id("qualitySnapshots"),
        _creationTime: v.number(),
        deploymentKey: v.string(),
      }),
    ),
    model: v.union(
      v.null(),
      modelProjection.extend({
        _id: v.id("modelProfiles"),
        _creationTime: v.number(),
        deploymentKey: v.string(),
      }),
    ),
  }),
  handler: async (ctx) => {
    // Only the synthetic preview is intentionally public. Real deployment
    // selection stays unavailable until tenant auth/authorization is wired.
    const deploymentKey = DEMO_DEPLOYMENT;
    const [deployment, repositories, reviews, quality, model] =
      await Promise.all([
        ctx.db
          .query("deployments")
          .withIndex("by_deployment_key", (q) =>
            q.eq("deploymentKey", deploymentKey),
          )
          .unique(),
        ctx.db
          .query("repositories")
          .withIndex("by_deployment_key_and_updated_at", (q) =>
            q.eq("deploymentKey", deploymentKey),
          )
          .order("desc")
          .take(50),
        ctx.db
          .query("reviewRuns")
          .withIndex("by_deployment_key_and_updated_at", (q) =>
            q.eq("deploymentKey", deploymentKey),
          )
          .order("desc")
          .take(40),
        ctx.db
          .query("qualitySnapshots")
          .withIndex("by_deployment_key_and_evaluated_at", (q) =>
            q.eq("deploymentKey", deploymentKey),
          )
          .order("desc")
          .first(),
        ctx.db
          .query("modelProfiles")
          .withIndex("by_deployment_key", (q) =>
            q.eq("deploymentKey", deploymentKey),
          )
          .unique(),
      ]);
    return { deployment, repositories, reviews, quality, model };
  },
});

// Non-sensitive sample projections for the development workspace. Production
// projections never contain source, diffs, prompts, or finding bodies.
export const bootstrapDemo = internalMutation({
  args: {},
  returns: v.object({ created: v.boolean() }),
  handler: async (ctx) => {
    const existing = await ctx.db
      .query("deployments")
      .withIndex("by_deployment_key", (q) =>
        q.eq("deploymentKey", DEMO_DEPLOYMENT),
      )
      .unique();
    if (existing) return { created: false };

    const now = Date.now();
    await ctx.db.insert("deployments", {
      deploymentKey: DEMO_DEPLOYMENT,
      name: "Diffuse Osaka",
      kind: "self_hosted",
      region: "Local Docker",
      status: "healthy",
      version: "0.1.0-dev",
      updatedAt: now,
    });

    const repositories = [
      {
        externalId: "diffuse",
        name: "diffuse/diffuse",
        provider: "github" as const,
        defaultBranch: "main",
        indexStatus: "ready" as const,
        reviewStatus: "idle" as const,
        lastIndexedAt: now - 4 * 60_000,
        lastReviewAt: now - 18 * 60_000,
        openFindingCount: 3,
        criticalFindingCount: 0,
      },
      {
        externalId: "api",
        name: "diffuse/api",
        provider: "github" as const,
        defaultBranch: "main",
        indexStatus: "indexing" as const,
        reviewStatus: "reviewing" as const,
        lastIndexedAt: now - 2 * 60 * 60_000,
        lastReviewAt: now - 8 * 60_000,
        openFindingCount: 1,
        criticalFindingCount: 1,
      },
      {
        externalId: "docs",
        name: "diffuse/docs",
        provider: "gitlab" as const,
        defaultBranch: "main",
        indexStatus: "stale" as const,
        reviewStatus: "attention" as const,
        lastIndexedAt: now - 26 * 60 * 60_000,
        lastReviewAt: now - 8 * 60 * 60_000,
        openFindingCount: 0,
        criticalFindingCount: 0,
      },
    ];
    for (const repository of repositories) {
      await ctx.db.insert("repositories", {
        deploymentKey: DEMO_DEPLOYMENT,
        ...repository,
        updatedAt: now,
      });
    }

    const reviews = [
      {
        externalId: "review-184",
        repositoryExternalId: "api",
        repositoryName: "diffuse/api",
        number: 184,
        title: "Harden webhook signature validation",
        status: "reviewing" as const,
        findingCount: 1,
        criticalCount: 1,
        model: "openai/gpt-4.1-mini",
        updatedAt: now - 36_000,
      },
      {
        externalId: "review-182",
        repositoryExternalId: "diffuse",
        repositoryName: "diffuse/diffuse",
        number: 182,
        title: "Add self-host runtime preflight",
        status: "published" as const,
        findingCount: 3,
        criticalCount: 0,
        latencyMs: 44_200,
        model: "openai/gpt-4.1-mini",
        updatedAt: now - 18 * 60_000,
      },
      {
        externalId: "review-177",
        repositoryExternalId: "docs",
        repositoryName: "diffuse/docs",
        number: 177,
        title: "Document GitLab installation flow",
        status: "failed" as const,
        findingCount: 0,
        criticalCount: 0,
        latencyMs: 9_100,
        model: "openai/gpt-4.1-mini",
        updatedAt: now - 8 * 60 * 60_000,
      },
    ];
    for (const review of reviews) {
      await ctx.db.insert("reviewRuns", {
        deploymentKey: DEMO_DEPLOYMENT,
        ...review,
      });
    }

    await ctx.db.insert("qualitySnapshots", {
      deploymentKey: DEMO_DEPLOYMENT,
      evaluatedAt: now,
      precision: 0.86,
      recall: 0.74,
      f1: 0.795,
      truePositives: 37,
      falsePositives: 6,
      falseNegatives: 13,
      addressedFindings: 21,
      sampleSize: 56,
      medianLatencyMs: 44_200,
      estimatedCostUsd: 0.12,
    });
    await ctx.db.insert("modelProfiles", {
      deploymentKey: DEMO_DEPLOYMENT,
      provider: "OpenAI",
      model: "openai/gpt-4.1-mini",
      status: "missing_key",
      reviewPasses: ["correctness", "security", "performance", "tests"],
      updatedAt: now,
    });
    return { created: true };
  },
});

export const ingestSnapshot = internalMutation({
  args: {
    deployment: deploymentProjection,
    repositories: v.array(repositoryProjection),
    reviews: v.array(reviewProjection),
    quality: v.optional(qualityProjection),
    model: v.optional(modelProjection),
  },
  returns: v.object({ accepted: v.boolean() }),
  handler: async (ctx, args) => {
    const deploymentKey = args.deployment.deploymentKey;
    const existingDeployment = await ctx.db
      .query("deployments")
      .withIndex("by_deployment_key", (q) =>
        q.eq("deploymentKey", deploymentKey),
      )
      .unique();
    if (existingDeployment) {
      await ctx.db.patch(existingDeployment._id, args.deployment);
    } else {
      await ctx.db.insert("deployments", args.deployment);
    }

    for (const repository of args.repositories) {
      const existing = await ctx.db
        .query("repositories")
        .withIndex("by_deployment_key_and_external_id", (q) =>
          q
            .eq("deploymentKey", deploymentKey)
            .eq("externalId", repository.externalId),
        )
        .unique();
      const value = { deploymentKey, ...repository };
      if (existing) await ctx.db.patch(existing._id, value);
      else await ctx.db.insert("repositories", value);
    }

    for (const review of args.reviews) {
      const existing = await ctx.db
        .query("reviewRuns")
        .withIndex("by_deployment_key_and_external_id", (q) =>
          q
            .eq("deploymentKey", deploymentKey)
            .eq("externalId", review.externalId),
        )
        .unique();
      const value = { deploymentKey, ...review };
      if (existing) await ctx.db.patch(existing._id, value);
      else await ctx.db.insert("reviewRuns", value);
    }

    if (args.quality) {
      const existing = await ctx.db
        .query("qualitySnapshots")
        .withIndex("by_deployment_key_and_evaluated_at", (q) =>
          q
            .eq("deploymentKey", deploymentKey)
            .eq("evaluatedAt", args.quality!.evaluatedAt),
        )
        .unique();
      const value = { deploymentKey, ...args.quality };
      if (existing) await ctx.db.patch(existing._id, value);
      else await ctx.db.insert("qualitySnapshots", value);
    }
    if (args.model) {
      const existing = await ctx.db
        .query("modelProfiles")
        .withIndex("by_deployment_key", (q) =>
          q.eq("deploymentKey", deploymentKey),
        )
        .unique();
      const value = { deploymentKey, ...args.model };
      if (existing) await ctx.db.patch(existing._id, value);
      else await ctx.db.insert("modelProfiles", value);
    }
    return { accepted: true };
  },
});
